src/data/dataset.py

```diff
@@ -19,88 +49,149 @@ class Dataset(object):
+            artists_in_split = []
+            for line in self.metadata.lines('%s.csv' % split):
+                artists_in_split.append(line.rstrip('\n'))
         else:
             dirs = []
+            all_artists = []
+            skipped_count = 0
+            last_log = 0
+            last_log_percent = None
+            pool = multiprocessing.Pool(multiprocessing.cpu_count())
+
             for artist in os.listdir(root):
                 if os.path.isdir(os.path.join(root, artist)):
-                    dirs.append(artist)
-            artists = []
-            skipped_count = 0
+                    songs = os.listdir(os.path.join(root, artist))
+                    songs = [song for song in songs if loader.is_song(song)]
+                    if len(songs) > 0:
+                        dirs.append(artist)
+
             num_dirs = len(dirs)
-            last_log = 0
+
             for artist_index, artist in enumerate(dirs):
-                # log progress every second
-                if time.time() - last_log >= 1:
-                    log.info("Preprocessing data. %s%%" % int(100*artist_index/num_dirs))
-                    last_log = time.time()
                 songs = os.listdir(os.path.join(root, artist))
-                if len(songs) >= min_songs:
-                    artists.append(artist)
+                # We only want .txt and .mid files. Filter all others.
+                songs = [song for song in songs if loader.is_song(song)]
+                # populate `valid_songs[artist]`
+                if validate:
+                    # log progress at most every second
+                    if time.time() - last_log >= 1:
+                        if last_log_percent != '%.2f' % (100*artist_index/num_dirs):
+                            last_log_percent = '%.2f' % (100*artist_index/num_dirs)
+                            log.info("Preprocessing data. %s%%" % last_log_percent)
+                            last_log = time.time()
+                    if artist not in valid_songs:
+                        valid_songs[artist] = set()
+                    songs_to_validate = [song for song in songs if song not in valid_songs[artist]]
+                    song_files = [os.path.join(root, artist, song) for song in songs_to_validate]
+                    if parallel:
+                        mapped = pool.map(loader.validate, song_files)
+                    else:
+                        mapped = map(loader.validate, song_files)
+                    validated = itertools.compress(songs_to_validate, mapped)
+                    for song in validated:
+                        song_file = os.path.join(root, artist, song)
+                        if persist:
+                            self.metadata.write('valid_songs.csv', '%s,%s\n' % (artist, song))
+                        valid_songs[artist].add(song)
+                else:
+                    valid_songs[artist] = set(songs)
```

Comment:
The reason why it's this way is because `valid_songs[artist]` might be partially populated from the file. So, if we're using persistence, we want to valid songs which haven't already been validated (i.e. aren't in the file) and then add them to the file.

Reply:
Thoughts?